scipy/sparse/csgraph/_flow.pyx

```diff
@@ -541,27 +542,60 @@ cdef ITYPE_t _augment_paths(
         An arbitrary flow in the layered graph.
 
     """
-    if current == sink:
-        return flow
-
-    cdef ITYPE_t dst_vertex, result_flow, e
-
-    while progress[current] < edge_ptr[current + 1]:
+    cdef ITYPE_t top, dst_vertex, result_flow
+    cdef ITYPE_t stack_top, e, current
+    top = 0
+    stack[top][0] = source
+    stack[top][1] = flow
+
+    while top > -1:
+        current = stack[top][0]
+        flow = stack[top][1]
         e = progress[current]
         dst_vertex = heads[e]
         if (capacities[e] > 0 and
                 levels[dst_vertex] == levels[current] + 1):
-            result_flow = _augment_paths(edge_ptr, dst_vertex, sink,
-                                         levels, heads, rev_edge_ptr,
-                                         capacities, progress, flows,
-                                         min(flow, capacities[e]))
-            if result_flow:
-                capacities[e] -= result_flow
-                capacities[rev_edge_ptr[e]] += result_flow
-                flows[e] += result_flow
-                flows[rev_edge_ptr[e]] -= result_flow
-                return result_flow
-        progress[current] += 1
+            if dst_vertex == sink:
+                result_flow = min(flow, capacities[e])
+                if result_flow > 0:
+                    while top > -1:
+                        e = progress[stack[top][0]]
+                        capacities[e] -= result_flow
+                        capacities[rev_edge_ptr[e]] += result_flow
+                        flows[e] += result_flow
+                        flows[rev_edge_ptr[e]] -= result_flow
+                        top -= 1
+                    return result_flow
+                while top > -1:
+                    if progress[stack[top][0]] < edge_ptr[stack[top][0] + 1]:
+                        progress[stack[top][0]] += 1
+                    if progress[stack[top][0]] >= edge_ptr[stack[top][0] + 1]:
+                        top -= 1
+                    else:
+                        break
+            else:
+                if progress[dst_vertex] >= edge_ptr[dst_vertex + 1]:
+                    while top > -1:
+                        stack_top = stack[top][0]
+                        if progress[stack_top] < edge_ptr[stack_top + 1]:
+                            progress[stack_top] += 1
+                        if progress[stack_top] >= edge_ptr[stack_top + 1]:
+                            top -= 1
+                        else:
+                            break
+                else:
+                    top += 1
+                    stack[top][0] = dst_vertex
+                    stack[top][1] = min(flow, capacities[e])
+        else:
+            while top > -1:
```

Comment:
It feels like it should only be necessary to unwind the stack in one place, where we currently have three. One looks like dead code, so that's easy enough; the two other ones seem like they could be merged if we just pop `(current, flow)` instead of keeping them on the stack, hence removing the need to manually unwind at all. Then again, they're pretty useful to have on the stack when you're augmenting (not having them around during augmentation is what caused the need for `tails` in `_edmonds_karp`). Hm.

Reply:
I played around with it a bit and also decided that explicitly popping, and keeping vertices on the stack, made for code that was easier to follow, but I think I managed to get rid of the duplication; see the diff in the main thread.